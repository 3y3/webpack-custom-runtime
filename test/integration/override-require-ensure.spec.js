'use strict';

const path = require('path');
const diff = require('diff');
const webpack = require('./utils/webpack');
const WebpackCustomRuntime = require('../../').CustomRuntimePlugin;

const mainOutputDir = path.join(__dirname, 'fixtures', 'dist');

describe.only('require-ensure override', function() {

    function cit(name, options, config) {
        test(name, async () => {
            const base = webpack();
            const curr = webpack(Object.assign({
                plugins: [
                    new WebpackCustomRuntime(options)
                ]
            }, config));

            await Promise.all([ base.result, curr.result ]);

            const baseDir = base.fs.readdirSync(mainOutputDir);
            const currDir = curr.fs.readdirSync(mainOutputDir);

            expect(baseDir).toEqual(currDir);

            currDir.forEach((file) => {
                const linesDiff = diff.createTwoFilesPatch(
                    file, file,
                    base.fs.readFileSync(mainOutputDir + '/' + file).toString(),
                    curr.fs.readFileSync(mainOutputDir + '/' + file).toString(),
                    '', '',
                    { context: 0 }
                );

                expect(linesDiff).toMatchSnapshot(file);
            });
        });
    }

    cit('should implement default webpack behavior', {}, {
        output: {
            path: mainOutputDir,
            filename: '[name].js',
            jsonpScriptType: 'module',
            chunkLoadTimeout: 60000,
            crossOriginLoading: 'anonymous',
        }
    });

    cit('should skip defaults on custom behavior', { behavior: '' });
});
