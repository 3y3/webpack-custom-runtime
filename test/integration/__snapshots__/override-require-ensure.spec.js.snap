// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`require-ensure override should implement default webpack behavior: 0.js 1`] = `
"Index: 0.js
===================================================================
--- 0.js	
+++ 0.js	
"
`;

exports[`require-ensure override should implement default webpack behavior: 1.js 1`] = `
"Index: 1.js
===================================================================
--- 1.js	
+++ 1.js	
"
`;

exports[`require-ensure override should implement default webpack behavior: main-module.js 1`] = `
"Index: main-module.js
===================================================================
--- main-module.js	
+++ main-module.js	
"
`;

exports[`require-ensure override should implement default webpack behavior: other-module.js 1`] = `
"Index: other-module.js
===================================================================
--- other-module.js	
+++ other-module.js	
"
`;

exports[`require-ensure override should implement default webpack behavior: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ -55,1 +55,1 @@
-/******/
+/******/ 					link.crossOrigin = \\"anonymous\\";
@@ -86,0 +86,2 @@
+/******/ 	function scriptUrlResolver(installedChunks, chunkId, mode) {
+/******/ 	  var result = (originalUrl = jsonpScriptSrc(chunkId));
@@ -87,0 +89,105 @@
+/******/ 	  mode = mode || 'load';
+/******/
+/******/ 	  return result;
+/******/ 	}
+/******/
+/******/ 	function scriptOptionsResolver(installedChunks, chunkId, url, mode) {
+/******/ 	  var result = {};
+/******/
+/******/ 	  mode = mode || 'load';
+/******/ 	  result.type = 'module';
+/******/ 	  result.timeout = 60;
+/******/ 	  if (__webpack_require__.nc) {
+/******/ 	    result.nonce = __webpack_require__.nc;
+/******/ 	  }
+/******/ 	  if (url.indexOf(window.location.origin + '/') !== 0) {
+/******/ 	    result.crossOrigin = 'anonymous';
+/******/ 	  }
+/******/
+/******/ 	  return result;
+/******/ 	}
+/******/
+/******/ 	function scriptBuilder(src, options, onSuccess, onError) {
+/******/ 	  var script, timeout;
+/******/
+/******/ 	  var onScriptComplete = function(error) {
+/******/ 	    onScriptComplete = function() {}; // avoid mem leaks in IE.
+/******/
+/******/ 	    script.onerror = script.onload = script.onreadystatechange = null;
+/******/ 	    clearTimeout(timeout);
+/******/ 	    error ? onError(error) : onSuccess();
+/******/ 	  };
+/******/
+/******/ 	  try {
+/******/ 	    options = extend(
+/******/ 	      {
+/******/ 	        // most compatible type
+/******/ 	        type: 'text/javascript',
+/******/ 	        charset: 'utf-8',
+/******/ 	        timeout: 20,
+/******/ 	        async: true
+/******/ 	      },
+/******/ 	      options
+/******/ 	    );
+/******/ 	    script = document.createElement('script');
+/******/ 	    extend(script, options);
+/******/
+/******/ 	    script.onload = function() {
+/******/ 	      onScriptComplete();
+/******/ 	    };
+/******/
+/******/ 	    timeout = setTimeout(function() {
+/******/ 	      onScriptComplete(error('timeout'));
+/******/ 	    }, options.timeout * 1000);
+/******/
+/******/ 	    script.onerror = function(event) {
+/******/ 	      onScriptComplete(
+/******/ 	        error(event && (event.type === 'load' ? 'missing' : event.type))
+/******/ 	      );
+/******/ 	    };
+/******/
+/******/ 	    script.onreadystatechange = function() {
+/******/ 	      // for compatibility with old IE
+/******/ 	      if (script.readyState === 'complete' || script.readyState === 'loaded') {
+/******/ 	        onScriptComplete(); // there is no way to catch loading errors in IE8
+/******/ 	      }
+/******/ 	    };
+/******/
+/******/ 	    script.src = src;
+/******/ 	  } catch (e) {
+/******/ 	    clearTimeout(timeout);
+/******/ 	    timeout = setTimeout(onScriptComplete, 0, error('script'));
+/******/ 	  }
+/******/
+/******/ 	  return script;
+/******/
+/******/ 	  function error(reason) {
+/******/ 	    var error = new Error('Failed to load resource');
+/******/ 	    error.type = reason;
+/******/ 	    return error;
+/******/ 	  }
+/******/
+/******/ 	  function extend(acc, src) {
+/******/ 	    if (src) {
+/******/ 	      for (var key in src) {
+/******/ 	        if (src.hasOwnProperty(key)) {
+/******/ 	          acc[key] = src[key];
+/******/ 	        }
+/******/ 	      }
+/******/ 	    }
+/******/
+/******/ 	    return acc;
+/******/ 	  }
+/******/ 	}
+/******/
+/******/ 	function scriptLoadHandler(installedChunks, chunkId) {
+/******/ 	  var result = null;
+/******/
+/******/ 	  return Promise.resolve(result);
+/******/ 	}
+/******/
+/******/ 	function scriptErrorHandler(installedChunks, chunkId, originalError) {
+/******/ 	  return Promise.reject(originalError);
+/******/ 	}
+/******/
+/******/
@@ -135,1 +242,1 @@
-/******/
+/******/ 				script.type = \\"module\\";
@@ -137,1 +244,1 @@
-/******/ 				script.timeout = 120;
+/******/ 				script.timeout = 60;
@@ -142,1 +249,3 @@
-/******/
+/******/ 				if (script.src.indexOf(window.location.origin + '/') !== 0) {
+/******/ 					script.crossOrigin = \\"anonymous\\";
+/******/ 				}
@@ -165,1 +274,1 @@
-/******/ 				}, 120000);
+/******/ 				}, 60000);
@@ -171,0 +280,118 @@
+/******/ 		var result = (function(
+/******/ 		  installedChunks,
+/******/ 		  chunkId,
+/******/ 		  scriptUrlResolver,
+/******/ 		  scriptOptionsResolver,
+/******/ 		  scriptBuilder,
+/******/ 		  scriptLoadHandler,
+/******/ 		  scriptErrorHandler
+/******/ 		) {
+/******/ 		  // JSONP chunk loading for javascript
+/******/ 		  var INSTALLED = 0;
+/******/ 		  var PROMISE_RESOLVE = 0;
+/******/ 		  var PROMISE_REJECT = 1;
+/******/ 		  var LOAD_PROMISE = 2;
+/******/ 		  var installedChunkData = installedChunks[chunkId];
+/******/
+/******/ 		  if (installedChunkData === INSTALLED || installedChunkData) {
+/******/ 		    return installedChunkData === INSTALLED
+/******/ 		      ? Promise.resolve()
+/******/ 		      : installedChunkData[LOAD_PROMISE];
+/******/ 		  }
+/******/
+/******/ 		  var url, options, script, error;
+/******/ 		  var currentStackError = new Error(); // create error before stack unwound to get useful stacktrace later
+/******/
+/******/ 		  installedChunkData = installedChunks[chunkId] = [];
+/******/ 		  installedChunkData[LOAD_PROMISE] = new Promise(function(resolve, reject) {
+/******/ 		    installedChunkData[PROMISE_RESOLVE] = resolve;
+/******/ 		    installedChunkData[PROMISE_REJECT] = reject;
+/******/ 		  })
+/******/ 		    .catch(setError)
+/******/ 		    .then(checkScriptLoaded)
+/******/ 		    .then(onLoad)
+/******/ 		    .catch(onError);
+/******/ 		  Promise.resolve()
+/******/ 		    .then(resolveUrlAndOptions)
+/******/ 		    .then(buildScript)
+/******/ 		    .then(setScript)
+/******/ 		    .then(appendScript)
+/******/ 		    .catch(installedChunkData[PROMISE_REJECT]);
+/******/ 		  return installedChunkData[LOAD_PROMISE];
+/******/
+/******/ 		  function checkScriptLoaded() {
+/******/ 		    var chunk = installedChunks[chunkId];
+/******/ 		    var isChunkComplete = chunk === INSTALLED; // ignore any errors if chunk is loaded
+/******/
+/******/ 		    if (isChunkComplete) {
+/******/ 		      return;
+/******/ 		    }
+/******/
+/******/ 		    installedChunks[chunkId] = null; // not 'undefined' for prevent useless prefetching
+/******/
+/******/ 		    error = error || currentStackError;
+/******/ 		    error.type = error.type || 'missing';
+/******/ 		    error.name = 'ChunkLoadError';
+/******/ 		    error.message =
+/******/ 		      'Loading chunk ' +
+/******/ 		      chunkId +
+/******/ 		      ' failed.\\\\n(' +
+/******/ 		      error.type +
+/******/ 		      ': ' +
+/******/ 		      script.src +
+/******/ 		      ')';
+/******/ 		    error.request = script.src; // Extra fields
+/******/
+/******/ 		    error.chunkId = chunkId;
+/******/ 		    throw error;
+/******/ 		  }
+/******/
+/******/ 		  function buildScript() {
+/******/ 		    // scriptBuilder can be asynchronous for old school strategies like \`blob loading\`
+/******/ 		    return scriptBuilder(
+/******/ 		      url,
+/******/ 		      options,
+/******/ 		      installedChunkData[PROMISE_RESOLVE],
+/******/ 		      installedChunkData[PROMISE_REJECT]
+/******/ 		    );
+/******/ 		  }
+/******/
+/******/ 		  function resolveUrlAndOptions() {
+/******/ 		    url = scriptUrlResolver(installedChunks, chunkId, 'load');
+/******/ 		    options = scriptOptionsResolver(installedChunks, chunkId, url, 'load');
+/******/ 		  }
+/******/
+/******/ 		  function appendScript() {
+/******/ 		    document.getElementsByTagName('head')[0].appendChild(script);
+/******/ 		  }
+/******/
+/******/ 		  function setError(_error) {
+/******/ 		    error = _error;
+/******/ 		  }
+/******/
+/******/ 		  function setScript(_script) {
+/******/ 		    script = _script;
+/******/ 		  }
+/******/
+/******/ 		  function onLoad() {
+/******/ 		    return scriptLoadHandler(installedChunks, chunkId);
+/******/ 		  }
+/******/
+/******/ 		  function onError(error) {
+/******/ 		    return scriptErrorHandler(installedChunks, chunkId, error);
+/******/ 		  }
+/******/ 		})(
+/******/ 		  installedChunks,
+/******/ 		  chunkId,
+/******/ 		  scriptUrlResolver,
+/******/ 		  scriptOptionsResolver,
+/******/ 		  scriptBuilder,
+/******/ 		  scriptLoadHandler,
+/******/ 		  scriptErrorHandler
+/******/ 		);
+/******/
+/******/ 		if (result) {
+/******/ 		  promises.push.apply(promises, [].concat(result));
+/******/ 		}
+/******/
+/******/
@@ -185,1 +412,1 @@
-/******/
+/******/ 					link.type = \\"module\\";
@@ -193,0 +420,3 @@
+/******/ 					if (link.href.indexOf(window.location.origin + '/') !== 0) {
+/******/ 						link.crossOrigin = \\"anonymous\\";
+/******/ 					}
"
`;

exports[`require-ensure override should implement default webpack behavior: test1.js 1`] = `
"Index: test1.js
===================================================================
--- test1.js	
+++ test1.js	
"
`;

exports[`require-ensure override should implement default webpack behavior: test2.js 1`] = `
"Index: test2.js
===================================================================
--- test2.js	
+++ test2.js	
"
`;

exports[`require-ensure override should implement default webpack behavior: test5.js 1`] = `
"Index: test5.js
===================================================================
--- test5.js	
+++ test5.js	
"
`;

exports[`require-ensure override should skip defaults on custom behavior: 0.js 1`] = `
"Index: 0.js
===================================================================
--- 0.js	
+++ 0.js	
"
`;

exports[`require-ensure override should skip defaults on custom behavior: 1.js 1`] = `
"Index: 1.js
===================================================================
--- 1.js	
+++ 1.js	
"
`;

exports[`require-ensure override should skip defaults on custom behavior: main-module.js 1`] = `
"Index: main-module.js
===================================================================
--- main-module.js	
+++ main-module.js	
"
`;

exports[`require-ensure override should skip defaults on custom behavior: other-module.js 1`] = `
"Index: other-module.js
===================================================================
--- other-module.js	
+++ other-module.js	
"
`;

exports[`require-ensure override should skip defaults on custom behavior: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ -86,0 +86,2 @@
+/******/ 	function scriptUrlResolver(installedChunks, chunkId, mode) {
+/******/ 	  var result = (originalUrl = jsonpScriptSrc(chunkId));
@@ -87,0 +89,97 @@
+/******/ 	  mode = mode || 'load';
+/******/
+/******/ 	  return result;
+/******/ 	}
+/******/
+/******/ 	function scriptOptionsResolver(installedChunks, chunkId, url, mode) {
+/******/ 	  var result = {};
+/******/
+/******/ 	  mode = mode || 'load';
+/******/
+/******/ 	  return result;
+/******/ 	}
+/******/
+/******/ 	function scriptBuilder(src, options, onSuccess, onError) {
+/******/ 	  var script, timeout;
+/******/
+/******/ 	  var onScriptComplete = function(error) {
+/******/ 	    onScriptComplete = function() {}; // avoid mem leaks in IE.
+/******/
+/******/ 	    script.onerror = script.onload = script.onreadystatechange = null;
+/******/ 	    clearTimeout(timeout);
+/******/ 	    error ? onError(error) : onSuccess();
+/******/ 	  };
+/******/
+/******/ 	  try {
+/******/ 	    options = extend(
+/******/ 	      {
+/******/ 	        // most compatible type
+/******/ 	        type: 'text/javascript',
+/******/ 	        charset: 'utf-8',
+/******/ 	        timeout: 20,
+/******/ 	        async: true
+/******/ 	      },
+/******/ 	      options
+/******/ 	    );
+/******/ 	    script = document.createElement('script');
+/******/ 	    extend(script, options);
+/******/
+/******/ 	    script.onload = function() {
+/******/ 	      onScriptComplete();
+/******/ 	    };
+/******/
+/******/ 	    timeout = setTimeout(function() {
+/******/ 	      onScriptComplete(error('timeout'));
+/******/ 	    }, options.timeout * 1000);
+/******/
+/******/ 	    script.onerror = function(event) {
+/******/ 	      onScriptComplete(
+/******/ 	        error(event && (event.type === 'load' ? 'missing' : event.type))
+/******/ 	      );
+/******/ 	    };
+/******/
+/******/ 	    script.onreadystatechange = function() {
+/******/ 	      // for compatibility with old IE
+/******/ 	      if (script.readyState === 'complete' || script.readyState === 'loaded') {
+/******/ 	        onScriptComplete(); // there is no way to catch loading errors in IE8
+/******/ 	      }
+/******/ 	    };
+/******/
+/******/ 	    script.src = src;
+/******/ 	  } catch (e) {
+/******/ 	    clearTimeout(timeout);
+/******/ 	    timeout = setTimeout(onScriptComplete, 0, error('script'));
+/******/ 	  }
+/******/
+/******/ 	  return script;
+/******/
+/******/ 	  function error(reason) {
+/******/ 	    var error = new Error('Failed to load resource');
+/******/ 	    error.type = reason;
+/******/ 	    return error;
+/******/ 	  }
+/******/
+/******/ 	  function extend(acc, src) {
+/******/ 	    if (src) {
+/******/ 	      for (var key in src) {
+/******/ 	        if (src.hasOwnProperty(key)) {
+/******/ 	          acc[key] = src[key];
+/******/ 	        }
+/******/ 	      }
+/******/ 	    }
+/******/
+/******/ 	    return acc;
+/******/ 	  }
+/******/ 	}
+/******/
+/******/ 	function scriptLoadHandler(installedChunks, chunkId) {
+/******/ 	  var result = null;
+/******/
+/******/ 	  return Promise.resolve(result);
+/******/ 	}
+/******/
+/******/ 	function scriptErrorHandler(installedChunks, chunkId, originalError) {
+/******/ 	  return Promise.reject(originalError);
+/******/ 	}
+/******/
+/******/
@@ -171,0 +270,118 @@
+/******/ 		var result = (function(
+/******/ 		  installedChunks,
+/******/ 		  chunkId,
+/******/ 		  scriptUrlResolver,
+/******/ 		  scriptOptionsResolver,
+/******/ 		  scriptBuilder,
+/******/ 		  scriptLoadHandler,
+/******/ 		  scriptErrorHandler
+/******/ 		) {
+/******/ 		  // JSONP chunk loading for javascript
+/******/ 		  var INSTALLED = 0;
+/******/ 		  var PROMISE_RESOLVE = 0;
+/******/ 		  var PROMISE_REJECT = 1;
+/******/ 		  var LOAD_PROMISE = 2;
+/******/ 		  var installedChunkData = installedChunks[chunkId];
+/******/
+/******/ 		  if (installedChunkData === INSTALLED || installedChunkData) {
+/******/ 		    return installedChunkData === INSTALLED
+/******/ 		      ? Promise.resolve()
+/******/ 		      : installedChunkData[LOAD_PROMISE];
+/******/ 		  }
+/******/
+/******/ 		  var url, options, script, error;
+/******/ 		  var currentStackError = new Error(); // create error before stack unwound to get useful stacktrace later
+/******/
+/******/ 		  installedChunkData = installedChunks[chunkId] = [];
+/******/ 		  installedChunkData[LOAD_PROMISE] = new Promise(function(resolve, reject) {
+/******/ 		    installedChunkData[PROMISE_RESOLVE] = resolve;
+/******/ 		    installedChunkData[PROMISE_REJECT] = reject;
+/******/ 		  })
+/******/ 		    .catch(setError)
+/******/ 		    .then(checkScriptLoaded)
+/******/ 		    .then(onLoad)
+/******/ 		    .catch(onError);
+/******/ 		  Promise.resolve()
+/******/ 		    .then(resolveUrlAndOptions)
+/******/ 		    .then(buildScript)
+/******/ 		    .then(setScript)
+/******/ 		    .then(appendScript)
+/******/ 		    .catch(installedChunkData[PROMISE_REJECT]);
+/******/ 		  return installedChunkData[LOAD_PROMISE];
+/******/
+/******/ 		  function checkScriptLoaded() {
+/******/ 		    var chunk = installedChunks[chunkId];
+/******/ 		    var isChunkComplete = chunk === INSTALLED; // ignore any errors if chunk is loaded
+/******/
+/******/ 		    if (isChunkComplete) {
+/******/ 		      return;
+/******/ 		    }
+/******/
+/******/ 		    installedChunks[chunkId] = null; // not 'undefined' for prevent useless prefetching
+/******/
+/******/ 		    error = error || currentStackError;
+/******/ 		    error.type = error.type || 'missing';
+/******/ 		    error.name = 'ChunkLoadError';
+/******/ 		    error.message =
+/******/ 		      'Loading chunk ' +
+/******/ 		      chunkId +
+/******/ 		      ' failed.\\\\n(' +
+/******/ 		      error.type +
+/******/ 		      ': ' +
+/******/ 		      script.src +
+/******/ 		      ')';
+/******/ 		    error.request = script.src; // Extra fields
+/******/
+/******/ 		    error.chunkId = chunkId;
+/******/ 		    throw error;
+/******/ 		  }
+/******/
+/******/ 		  function buildScript() {
+/******/ 		    // scriptBuilder can be asynchronous for old school strategies like \`blob loading\`
+/******/ 		    return scriptBuilder(
+/******/ 		      url,
+/******/ 		      options,
+/******/ 		      installedChunkData[PROMISE_RESOLVE],
+/******/ 		      installedChunkData[PROMISE_REJECT]
+/******/ 		    );
+/******/ 		  }
+/******/
+/******/ 		  function resolveUrlAndOptions() {
+/******/ 		    url = scriptUrlResolver(installedChunks, chunkId, 'load');
+/******/ 		    options = scriptOptionsResolver(installedChunks, chunkId, url, 'load');
+/******/ 		  }
+/******/
+/******/ 		  function appendScript() {
+/******/ 		    document.getElementsByTagName('head')[0].appendChild(script);
+/******/ 		  }
+/******/
+/******/ 		  function setError(_error) {
+/******/ 		    error = _error;
+/******/ 		  }
+/******/
+/******/ 		  function setScript(_script) {
+/******/ 		    script = _script;
+/******/ 		  }
+/******/
+/******/ 		  function onLoad() {
+/******/ 		    return scriptLoadHandler(installedChunks, chunkId);
+/******/ 		  }
+/******/
+/******/ 		  function onError(error) {
+/******/ 		    return scriptErrorHandler(installedChunks, chunkId, error);
+/******/ 		  }
+/******/ 		})(
+/******/ 		  installedChunks,
+/******/ 		  chunkId,
+/******/ 		  scriptUrlResolver,
+/******/ 		  scriptOptionsResolver,
+/******/ 		  scriptBuilder,
+/******/ 		  scriptLoadHandler,
+/******/ 		  scriptErrorHandler
+/******/ 		);
+/******/
+/******/ 		if (result) {
+/******/ 		  promises.push.apply(promises, [].concat(result));
+/******/ 		}
+/******/
+/******/
"
`;

exports[`require-ensure override should skip defaults on custom behavior: test1.js 1`] = `
"Index: test1.js
===================================================================
--- test1.js	
+++ test1.js	
"
`;

exports[`require-ensure override should skip defaults on custom behavior: test2.js 1`] = `
"Index: test2.js
===================================================================
--- test2.js	
+++ test2.js	
"
`;

exports[`require-ensure override should skip defaults on custom behavior: test5.js 1`] = `
"Index: test5.js
===================================================================
--- test5.js	
+++ test5.js	
"
`;
